L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorAutoEncoderImpl.cc

```diff
@@ -0,0 +1,193 @@
+    uint cellu = id.triggerCellU();
+    uint cellv = id.triggerCellV();
+    uint inputIndex = cellRemap_.at(cellu * 8 + cellv);
+    mipPt_[inputIndex] = trigCell.mipPt();
+    uncompressedCharge_[inputIndex] = trigCell.uncompressedCharge();
+    compressedCharge_[inputIndex] = trigCell.compressedCharge();
+
+    modSum += trigCell.mipPt();
+  }
+
+  if (modSum > 0) {
+    //normalize inputs to module sum
+    for (int i = 0; i < 48; i++) {
+      ae_inputArray[i] = mipPt_[i] / modSum;
+
+      //round to precision of input, if bitsPerInput_ is -1 keep full precision
+      if (bitsPerInput_ > 0) {
+        ae_inputArray[i] = round(ae_inputArray[i] * pow(2., bitsPerInput_)) / pow(2., bitsPerInput_);
+      }
+    }
+  }
+
+  tensorflow::Tensor encoder_input(tensorflow::DT_FLOAT, {1, 4, 4, 3});
+  float* d = encoder_input.flat<float>().data();
+  for (int i = 0; i < 48; i++, d++) {
+    *d = ae_inputArray[i];
+  }
+
+  int graphIndex = linkToGraphMap_.at(nLinks);
+
+  std::vector<tensorflow::Tensor> encoder_outputs;
+  tensorflow::run(session_encoder_.at(graphIndex),
+                  {{inputTensorName_encoder_, encoder_input}},
+                  {outputTensorName_encoder_},
+                  &encoder_outputs);
+
+  double ae_encodedLayer[16];
+  d = encoder_outputs[0].flat<float>().data();
+  for (int i = 0; i < encoder_outputs[0].NumElements(); i++, d++) {
+    ae_encodedLayer[i] = *d;
+    //truncate the encoded layer bits
+    if (bitsPerOutput > 0) {
+      ae_encodedLayer[i] = round(ae_encodedLayer[i] * pow(2, bitsPerOutput)) / pow(2., bitsPerOutput);
+    }
+  }
+
+  tensorflow::Tensor decoder_input(tensorflow::DT_FLOAT, {1, 16});
+  d = decoder_input.flat<float>().data();
+  for (int i = 0; i < 16; i++, d++) {
+    *d = ae_encodedLayer[i];
+  }
+
+  std::vector<tensorflow::Tensor> decoder_outputs;
+  tensorflow::run(session_decoder_.at(graphIndex),
+                  {{inputTensorName_decoder_, decoder_input}},
+                  {outputTensorName_decoder_},
+                  &decoder_outputs);
+
+  double ae_outputArray[48];
+
+  d = decoder_outputs[0].flat<float>().data();
+  for (int i = 0; i < decoder_outputs[0].NumElements(); i++, d++) {
+    ae_outputArray[i] = *d;
+  }
+
+  // Add data back into trigger cells
+  if (modSum > 0) {
+    //get detID for everything but cell, take first entry detID and subtract off cellU and cellV contribution
+    HGCalTriggerDetId id(trigCellVecInput.at(0).detId());
+    int cellU_ = id.triggerCellU();
+    int cellV_ = id.triggerCellV();
+    unsigned _id_waferBase = trigCellVecInput.at(0).detId() - (cellU_ << HGCalTriggerDetId::kHGCalCellUOffset) -
+                             (cellV_ << HGCalTriggerDetId::kHGCalCellVOffset);
+
+    //use first TC to find mipPt conversions to Et and ADC
+    float mipPtToEt_conv = trigCellVecInput.at(0).et() / trigCellVecInput.at(0).mipPt();
+    float mipPtToADC_conv = trigCellVecInput.at(0).hwPt() / trigCellVecInput.at(0).mipPt();
```

Comment:
The ADC conversion won't work, because it is actually the charge which is stored here (not the "transverse charge")

Reply:
I am curious why this would not work.  The output is intended to be the charge (similar to what is stored in output from the other algorithms), so the hwPt to mipPt ratio should give the conversion we want.

Reply:
The conversion factor is taken from the first TC. But this conversion depends on eta since the ADC count is not transverse while the mipPt is. So this factor cannot be applied on all TCs in the module later on.

Reply:
Thanks, I had somehow completely overlooked this fact.  I have updated this to add in a cosh(eta) factor to account for the transverse factor so it is MIP to ADC, and then when it's applied I similarly include the cosh(eta) factor for the trigger cell.

```
    float mipToADC_conv = trigCellVecInput[0].hwPt() / (trigCellVecInput[0].mipPt() * cosh(trigCellVecInput[0].eta()));
...
    for (int i = 0; i < nTriggerCells_; i++) {
...
        double mipPt = ae_outputArray[i] * modSum;
        double adc = mipPt * cosh(point.eta()) * mipToADC_conv;

```